pkg/action/executor/rescheduler.go

```diff
@@ -0,0 +1,261 @@
+
+	var machineIPs []string
+	switch newSEType {
+	case proto.EntityDTO_VIRTUAL_MACHINE:
+		// K8s uses Ip address as the Identifier. The VM name passed by actionItem is the display name
+		// that discovered by hypervisor. So here we must get the ip address from virtualMachineData in
+		// targetNode entityDTO.
+		vmData := targetNode.GetVirtualMachineData()
+		if vmData == nil {
+			return nil, errors.New("Missing VirtualMachineData in ActionItemDTO from server")
+		}
+		machineIPs = vmData.GetIpAddress()
+		break
+	case proto.EntityDTO_PHYSICAL_MACHINE:
+		// TODO
+		// machineIPS = <valid physical machine IP>
+		break
+	}
+	if machineIPs == nil || len(machineIPs) == 0 {
+		return nil, errors.New("Miss IP addresses in ActionItemDTO.")
+	}
+	glog.V(3).Infof("The IPs of targetNode is %v", machineIPs)
+
+	// Get the actual node name from Kubernetes cluster based on IP address.
+	nodeIdentifier, err := util.GetNodeNameFromIP(r.kubeClient, machineIPs)
+	if err != nil {
+		return nil, err
+	}
+
+	// Build action content.
+	var parentObjRef *turboaction.ParentObjectRef
+	// TODO, Ignore error?
+	parentRefObject, err := probe.FindParentReferenceObject(originalPod)
+	if err != nil {
+		return nil, err
+	}
+	if parentRefObject != nil {
+		parentObjRef = &turboaction.ParentObjectRef{
+			ParentObjectUID:       string(parentRefObject.UID),
+			ParentObjectNamespace: parentRefObject.Namespace,
+			ParentObjectName:      parentRefObject.Name,
+			ParentObjectType:      parentRefObject.Kind,
+		}
+	}
+	targetObj := &turboaction.TargetObject{
+		TargetObjectUID:       string(originalPod.UID),
+		TargetObjectNamespace: originalPod.Namespace,
+		TargetObjectName:      originalPod.Name,
+		TargetObjectType:      turboaction.TypePod,
+	}
+	moveSpec := turboaction.MoveSpec{
+		Source:      originalPod.Spec.NodeName,
+		Destination: nodeIdentifier,
+	}
+	content := turboaction.NewTurboActionContentBuilder(turboaction.ActionMove, targetObj).
+		ActionSpec(moveSpec).
+		ParentObjectRef(parentObjRef).
+		Build()
+
+	// Build TurboAction.
+	action := turboaction.NewTurboActionBuilder(originalPod.Namespace, *actionItem.Uuid).
+		Content(content).
+		Create()
+	return &action, nil
+}
+
+func (r *ReScheduler) reSchedule(action *turboaction.TurboAction) (*turboaction.TurboAction, error) {
+	// 1. Setup consumer
+	actionContent := action.Content
+	// Move destination should not be empty
+	moveSpec, ok := actionContent.ActionSpec.(turboaction.MoveSpec)
+	if !ok || moveSpec.Destination == "" {
+		return nil, errors.New("Failed to setup re-scheduler as move destination is nil after pre-precess.")
+	}
+
+	var key string
+	if actionContent.ParentObjectRef.ParentObjectUID != "" {
+		key = actionContent.ParentObjectRef.ParentObjectUID
+	} else if actionContent.TargetObject.TargetObjectUID != "" {
+		key = actionContent.TargetObject.TargetObjectUID
+	} else {
+		return nil, errors.New("Failed to setup re-scheduler consumer: failed to retrieve the key.")
```

Comment:
Again, key is the variable name, the error is about cannot find UID of the pod or provider